### PR TITLE
Lazy array display (#3638)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

The output for nested types has been made more consistent, fields are no longer enclosed in `"` without any escaping, and the same null sentinel is used as elsewhere.

Invalid timezones will result in an error, as opposed to rendering the error to the output string.

Date64 prints out the encoded time, previously it only printed the date which was incorrect.

### CODE REVIEW

File: arrow-cast/src/display.rs [hunk 1]
```diff
@@ -19,50 +19,301 @@
...
+impl<F: DisplayIndexState + Array> DisplayIndex for ArrayFormat<F> {
+    fn fmt(&self, idx: usize, f: &mut Formatter<'_>) -> FormatResult {
+        if self.array.is_null(idx) {
+            return Ok(());
+        }
+        DisplayIndexState::fmt(&self.array, &self.state, idx, f)
+    }
+}
+
+macro_rules! primitive_display {
```
Comment (tustvold) [inline arrow-cast/src/display.rs:169]: Unfortunately the lack of trait specialization means we are forced to use macros here

File: arrow-json/src/writer.rs [hunk 2]
```diff
@@ -315,47 +293,24 @@ fn set_column_for_json_rows(
...
-        }
-        DataType::Duration(TimeUnit::Nanosecond) => {
-            set_temporal_column_by_array_type!(col_name, col_idx, rows, array, row_count);
+        DataType::Date32
+        | DataType::Date64
+        | DataType::Timestamp(_, _)
+        | DataType::Time32(_)
+        | DataType::Time64(_)
+        | DataType::Duration(_) => {
+            // TODO: Set options
```
Comment (alamb) [inline arrow-json/src/writer.rs:302]: Is this a TODO for this PR or for some future PR? It seems like the default options work well enough for this?

File: arrow-json/src/writer.rs [hunk 3]
```diff
@@ -937,7 +892,7 @@ mod tests {
 
         assert_json_eq(
             &buf,
-            r#"{"date32":"2018-11-13","date64":"2018-11-13","name":"a"}
+            r#"{"date32":"2018-11-13","date64":"2018-11-13T17:11:10.011","name":"a"}
```
Comment (alamb) [inline arrow-json/src/writer.rs:895]: Is the idea for this change that since Date64 can repreent milliseconds it should be formatted like a timestamp even thought that may be surprising?

https://github.com/apache/arrow/blob/39bad5442c6447bf07594b09e4b29118b3211460/format/Schema.fbs#L214-L215

I admit the docs aren't very helpful in this respect: https://docs.rs/arrow/32.0.0/arrow/array/type.Date64Array.html

File: arrow-schema/src/datatype.rs [hunk 4]
```diff
@@ -290,7 +290,7 @@ pub enum IntervalUnit {
 }
 
 // Sparse or Dense union layouts
-#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Copy)]
```
Comment (alamb) [inline arrow-schema/src/datatype.rs:293]: 👍

File: arrow/src/util/pretty.rs [hunk 5]
```diff
@@ -447,48 +445,24 @@ mod tests {
...
-            "|                                                 |",
-            "+-------------------------------------------------+",
-        ];
-        check_datetime_with_timezone!(
-            TimestampSecondArray,
-            11111111,
-            "08:00".to_string(),
-            expected
-        );
+        let batch = timestamp_batch::<TimestampSecondType>("08:00", 11111111);
```
Comment (alamb) [inline arrow/src/util/pretty.rs:455]: I wonder why this change in behavior to error rather than print out "ERROR" -- is it because there was no way to return errors before? I wonder if people would be relying on the old behavior somehow?

File: arrow/src/util/pretty.rs [hunk 6]
```diff
@@ -559,12 +533,12 @@ mod tests {
...
-            "+------------+",
-            "| f          |",
-            "+------------+",
-            "| 2005-03-18 |",
-            "|            |",
-            "+------------+",
+            "+---------------------+",
+            "| f                   |",
+            "+---------------------+",
+            "| 2005-03-18T01:58:20 |",
```
Comment (alamb) [inline arrow/src/util/pretty.rs:539]: This is the same change as in the json code -- I realize Date64 has precision greater than a day, but I wonder if people will be surprised by this change 🤔

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 7]
```diff
@@ -2314,17 +2314,17 @@ mod tests {
...
-            | {"list": null}                                                                                                                      |
-            | {"list": [{"leaf_a": 2, "leaf_b": null}, {"leaf_a": 3, "leaf_b": 2}]}                                                               |
-            | {"list": null}                                                                                                                      |
-            | {"list": [{"leaf_a": 4, "leaf_b": null}, {"leaf_a": 5, "leaf_b": null}]}                                                            |
-            | {"list": [{"leaf_a": 6, "leaf_b": null}, {"leaf_a": 7, "leaf_b": null}, {"leaf_a": 8, "leaf_b": null}, {"leaf_a": 9, "leaf_b": 1}]} |
-            | {"list": [{"leaf_a": 10, "leaf_b": null}]}                                                                                          |
-            +-------------------------------------------------------------------------------------------------------------------------------------+
+            +-------------------------------------------------------------------------------------------------------+
```
Comment (alamb) [inline parquet/src/arrow/arrow_writer/mod.rs:2317]: this change certainly looks nicer to me -- I wonder if someone wants the old behavior (which looks like JSON), they should use the JSON writer, correct?

File: arrow-cast/src/display.rs [hunk 8]
```diff
@@ -19,57 +19,499 @@
...
+            datetime_format: None,
+            timestamp_format: None,
+            timestamp_tz_format: None,
+            time_format: None,
+        }
+    }
+}
+
+impl<'a> FormatOptions<'a> {
+    /// If set to `true` any formatting errors will be written to the output
```
Comment (alamb) [inline arrow-cast/src/display.rs:74]: 👍

File: arrow-cast/src/display.rs [hunk 9]
```diff
@@ -19,57 +19,499 @@
...
+    pub fn with_time_format(self, time_format: Option<&'a str>) -> Self {
+        Self {
+            time_format,
+            ..self
+        }
+    }
+}
+
+/// Implements [`Display`] for a specific array value
+pub struct ValueFormatter<'a> {
```
Comment (alamb) [inline arrow-cast/src/display.rs:130]: this is quite clever 👏

File: arrow-cast/src/display.rs [hunk 10]
```diff
@@ -19,57 +19,499 @@
...
+    idx: usize,
+    formatter: &'a ArrayFormatter<'a>,
+}
+
+impl<'a> ValueFormatter<'a> {
+    /// Writes this value to the provided [`Write`]
+    ///
+    /// Note: this ignores [`FormatOptions::with_display_error`]
```
Comment (alamb) [inline arrow-cast/src/display.rs:138]:
```suggestion
    /// Note: this ignores [`FormatOptions::with_display_error`] and
    /// will return an error on formatting issue
```

**Review by alamb — COMMENTED** [review]
In general I think this PR looks quite good and make the code cleaner and easier to understand. Thank you so much @tustvold 

I didn't quite get through the entire `display.rs` file yet but what I did get through is quite neat and looks good so far. 

I believe this is related to https://github.com/apache/arrow-rs/issues/3638

> Date64 prints out the encoded time, previously it only printed the date which was incorrect.

While I think I understand the rationale for this, I think it is worth double checking and maybe considering offering an escape valve formatting option if people want the old behavior.

File: arrow-cast/src/display.rs [hunk 11]
```diff
@@ -19,57 +19,499 @@
...
+impl<'a> DisplayIndex for &'a PrimitiveArray<Float16Type> {
+    fn write(&self, idx: usize, f: &mut dyn Write) -> FormatResult {
+        write!(f, "{}", self.value(idx))?;
+        Ok(())
+    }
+}
+
+macro_rules! decimal_display {
+    ($($t:ty),+) => {
+        $(impl<'a> DisplayIndexState<'a> for &'a PrimitiveArray<$t> {
```
Comment (alamb) [inline arrow-cast/src/display.rs:356]: I am impressed that you got the types to work out with references within macros 🤯

File: arrow-cast/src/display.rs [hunk 12]
```diff
@@ -135,657 +575,218 @@ macro_rules! make_string_interval_month_day_nano {
...
+        Ok(())
     }
+}
 
-    let dict_index = keys_array.value(row).as_usize();
-    array_value_to_string(dict_array.values(), dict_index)
+/// Get the value at the given row in an array as a String.
+///
+/// Note this function is quite inefficient and is unlikely to be
+/// suitable for converting large arrays or record batches.
```
Comment (alamb) [inline arrow-cast/src/display.rs:782]:
```suggestion
/// suitable for converting large arrays or record batches.
///
/// Please see [`ArrayFormatter`] for a better interface

```

File: arrow-cast/src/display.rs [hunk 13]
```diff
@@ -19,57 +19,499 @@
...
+            Err(FormatError::Arrow(e)) if self.formatter.safe => {
+                write!(f, "ERROR: {e}")
+            }
+            Err(_) => Err(std::fmt::Error),
+        }
+    }
+}
+
+/// A string formatter for an [`Array`]
+pub struct ArrayFormatter<'a> {
```
Comment (alamb) [inline arrow-cast/src/display.rs:170]: As far as I understand this is the main new interface. I think if we added an example to the docs here it would be great and help this usability a lot. A follow on PR would be fine (and I would be happy to help write it as well)

**Review by alamb — APPROVED** [review]
I finished reading this PR -- it is really nicely done @tustvold 🏅

File: arrow-json/src/writer.rs (hunk 3 above)
Comment (tustvold) [reply arrow-json/src/writer.rs:895]: Perhaps Date64 represents a DateTIme? It is a rather strange data type

File: parquet/src/arrow/arrow_writer/mod.rs (hunk 7 above)
Comment (tustvold) [reply parquet/src/arrow/arrow_writer/mod.rs:2317]: Yes, if someone wants actual JSON, they should use the JSON writer, what this previously produced was quasi-JSON that wouldn't illegal escape characters
